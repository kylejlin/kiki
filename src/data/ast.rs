#[derive(Clone, Debug)]
pub struct File {
    pub items: Vec<Item>,
}

#[derive(Clone, Debug)]
pub enum Item {
    Start(Ident),
    Struct(StructDef),
    Enum(EnumDef),
    Terminal(TerminalDef),
}

#[derive(Clone, Debug)]
pub struct StructDef {
    pub name: Ident,
    pub fieldset: Fieldset,
}

#[derive(Clone, Debug)]
pub struct EnumDef {
    pub name: Ident,
    pub variants: Vec<EnumVariant>,
}

#[derive(Clone, Debug)]
pub struct TerminalDef {
    pub name: Ident,
    pub variants: Vec<TerminalVariant>,
}

#[derive(Clone, Debug)]
pub enum Fieldset {
    Empty,
    Named(NamedFieldset),
    Tuple(TupleFieldset),
}

#[derive(Clone, Debug)]
pub struct NamedFieldset {
    pub fields: Vec<NamedField>,
}

#[derive(Clone, Debug)]
pub struct NamedField {
    pub name: IdentOrUnderscore,
    pub symbol: IdentOrTerminalIdent,
}

#[derive(Clone, Debug)]
pub struct TupleFieldset {
    pub fields: Vec<TupleField>,
}

#[derive(Clone, Debug)]
pub enum TupleField {
    Used(IdentOrTerminalIdent),
    Skipped(IdentOrTerminalIdent),
}

#[derive(Clone, Debug)]
pub struct EnumVariant {
    pub name: Ident,
    pub fieldset: Fieldset,
}

#[derive(Clone, Debug)]
pub struct TerminalVariant {
    pub name: TerminalIdent,
    pub type_: Type,
}

#[derive(Clone, Debug)]
pub enum Type {
    Unit,
    Path(Path),
    Complex(Box<ComplexType>),
}

#[derive(Clone, Debug)]
pub enum Path {
    One(Ident),
    Cons(Box<Path>, Ident),
}

#[derive(Clone, Debug)]
pub struct ComplexType {
    pub callee: Type,
    pub types: Vec<Type>,
}

pub use crate::data::cst::{Ident, IdentOrTerminalIdent, IdentOrUnderscore, TerminalIdent, Token};
