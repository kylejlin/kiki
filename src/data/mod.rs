pub mod ast;
pub mod cst;
pub mod machine;
pub mod table;
pub mod validated_file;

#[derive(Debug)]
pub enum KikiErr {
    Parse(ByteIndex, String, ByteIndex),
    NoStartSymbol,
    MultipleStartSymbols(Vec<ByteIndex>),
    NoTerminalEnum,
    MultipleTerminalEnums(Vec<ByteIndex>),
    SymbolsFirstLetteNotCapitalized(ByteIndex),
    DuplicateTerminalVariants(String, ByteIndex, ByteIndex),
    DuplicateNonterminals(String, ByteIndex, ByteIndex),
    UndefinedNonterminal(String, ByteIndex),
    UndefinedTerminal(DollarlessTerminalName, ByteIndex),
}

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Default)]
pub struct RustSrc(pub String);

#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord)]
pub struct ByteIndex(pub usize);

pub use crate::logic::oset::Oset;

use validated_file::DollarlessTerminalName;
