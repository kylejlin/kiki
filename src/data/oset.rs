#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct Oset<T> {
    raw: Vec<T>,
}

impl<T> Oset<T> {
    pub fn new() -> Self {
        Self { raw: Vec::new() }
    }
}

impl<T> Oset<T>
where
    T: Ord,
{
    pub fn insert(&mut self, item: T) {
        match self.raw.binary_search(&item) {
            Ok(_) => {}
            Err(i) => self.raw.insert(i, item),
        }
    }

    pub fn contains(&self, item: &T) -> bool {
        self.raw.binary_search(item).is_ok()
    }
}

impl<T> FromIterator<T> for Oset<T>
where
    T: Ord,
{
    fn from_iter<I>(iter: I) -> Self
    where
        I: IntoIterator<Item = T>,
    {
        let mut raw: Vec<T> = iter.into_iter().collect();
        raw.sort();
        raw.dedup();
        Self { raw }
    }
}

impl<T> IntoIterator for Oset<T> {
    type Item = T;
    type IntoIter = std::vec::IntoIter<T>;

    fn into_iter(self) -> Self::IntoIter {
        self.raw.into_iter()
    }
}
