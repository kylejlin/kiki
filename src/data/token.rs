pub use crate::data::*;

pub use crate::pipeline::parser::Token;

impl PartialEq for Token {
    fn eq(&self, other: &Self) -> bool {
        match (self, other) {
            (Token::Underscore(a), Token::Underscore(b)) => a == b,
            (Token::Ident(a), Token::Ident(b)) => a == b,
            (Token::TerminalIdent(a), Token::TerminalIdent(b)) => a == b,
            (Token::StartKw(a), Token::StartKw(b)) => a == b,
            (Token::StructKw(a), Token::StructKw(b)) => a == b,
            (Token::EnumKw(a), Token::EnumKw(b)) => a == b,
            (Token::TerminalKw(a), Token::TerminalKw(b)) => a == b,
            (Token::Colon(a), Token::Colon(b)) => a == b,
            (Token::DoubleColon(a), Token::DoubleColon(b)) => a == b,
            (Token::Comma(a), Token::Comma(b)) => a == b,
            (Token::LParen(a), Token::LParen(b)) => a == b,
            (Token::RParen(a), Token::RParen(b)) => a == b,
            (Token::LCurly(a), Token::LCurly(b)) => a == b,
            (Token::RCurly(a), Token::RCurly(b)) => a == b,
            (Token::LAngle(a), Token::LAngle(b)) => a == b,
            (Token::RAngle(a), Token::RAngle(b)) => a == b,
            _ => false,
        }
    }
}

impl Eq for Token {}

#[derive(Clone, Debug, PartialEq, Eq, Hash)]
pub struct Ident {
    pub name: String,
    pub position: ByteIndex,
}

#[derive(Clone, Debug, PartialEq, Eq, Hash)]
pub struct TerminalIdent {
    pub name: DollarlessTerminalName,
    pub dollarless_position: ByteIndex,
}

#[derive(Clone, Debug, PartialEq, Eq, Hash)]
pub struct Attribute {
    pub src: String,
    pub position: ByteIndex,
}
