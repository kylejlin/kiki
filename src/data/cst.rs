use crate::data::*;

#[derive(Clone, Debug)]
pub struct File {
    pub items: OptItems,
}

#[derive(Clone, Debug)]
pub enum OptItems {
    Nil,
    Cons(Box<OptItems>, FileItem),
}

#[derive(Clone, Debug)]
pub enum FileItem {
    Start(Ident),
    Struct(Struct),
    Enum(Enum),
    Terminal(TerminalEnum),
}

#[derive(Clone, Debug)]
pub struct Struct {
    pub name: Ident,
    pub fieldset: Fieldset,
}

#[derive(Clone, Debug)]
pub struct Enum {
    pub name: Ident,
    pub variants: OptEnumVariants,
}

#[derive(Clone, Debug)]
pub struct TerminalEnum {
    pub name: Ident,
    pub variants: OptTerminalEnumVariants,
}

#[derive(Clone, Debug)]
pub enum Fieldset {
    Empty,
    Named(NamedFieldset),
    Tuple(TupleFieldset),
}

#[derive(Clone, Debug)]
pub struct NamedFieldset {
    pub fields: NamedFields,
}

#[derive(Clone, Debug)]
pub enum NamedFields {
    One(NamedField),
    Cons(Box<NamedFields>, NamedField),
}

#[derive(Clone, Debug)]
pub struct NamedField {
    pub name: IdentOrUnderscore,
    pub symbol: IdentOrTerminalIdent,
}

#[derive(Clone, Debug)]
pub struct TupleFieldset {
    pub fields: TupleFields,
}

#[derive(Clone, Debug)]
pub enum TupleFields {
    One(TupleField),
    Cons(Box<TupleFields>, TupleField),
}

#[derive(Clone, Debug)]
pub enum TupleField {
    Used(IdentOrTerminalIdent),
    Skipped(IdentOrTerminalIdent),
}

#[derive(Clone, Debug)]
pub enum OptEnumVariants {
    Nil,
    Cons(Box<OptEnumVariants>, EnumVariant),
}

#[derive(Clone, Debug)]
pub struct EnumVariant {
    pub name: Ident,
    pub fieldset: Fieldset,
}

#[derive(Clone, Debug)]
pub enum OptTerminalEnumVariants {
    Nil,
    Cons(Box<OptTerminalEnumVariants>, TerminalEnumVariant),
}

#[derive(Clone, Debug)]
pub struct TerminalEnumVariant {
    pub name: TerminalIdent,
    pub type_: Type,
}

#[derive(Clone, Debug)]
pub enum Type {
    Unit,
    Path(Path),
    Complex(Box<ComplexType>),
}

#[derive(Clone, Debug)]
pub enum Path {
    One(Ident),
    Cons(Box<Path>, Ident),
}

#[derive(Clone, Debug)]
pub struct ComplexType {
    pub callee: Path,
    pub args: CommaSeparatedTypes,
}

#[derive(Clone, Debug)]
pub enum CommaSeparatedTypes {
    One(Type),
    Cons(Box<CommaSeparatedTypes>, Type),
}

#[derive(Clone, Debug)]
pub enum IdentOrUnderscore {
    Ident(Ident),
    Underscore,
}

#[derive(Clone, Debug)]
pub enum IdentOrTerminalIdent {
    Ident(Ident),
    Terminal(TerminalIdent),
}

#[derive(Clone, Debug)]
pub enum Token {
    Underscore,
    Ident(Ident),
    TerminalIdent(TerminalIdent),

    StartKw,
    StructKw,
    EnumKw,
    TerminalKw,

    Colon,
    DoubleColon,
    Comma,

    LParen,
    RParen,
    LCurly,
    RCurly,
    LAngle,
    RAngle,
}

#[derive(Clone, Debug)]
pub struct Ident {
    pub name: String,
    pub position: ByteIndex,
}

#[derive(Clone, Debug)]
pub struct TerminalIdent {
    pub name: DollarlessTerminalName,
    pub dollarless_position: ByteIndex,
}
