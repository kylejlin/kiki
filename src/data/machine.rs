use crate::data::{validated_file::DollarlessTerminalName, Oset};

#[derive(Debug, Clone)]
pub struct Machine {
    /// The first state is the start state.
    pub states: Vec<State>,
    pub transitions: Vec<Transition>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct State {
    pub items: Oset<Item>,
}

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct Item {
    pub rule_index: RuleIndex,
    pub lookahead: Lookahead,
    /// The `dot` is the index of the symbol to the right of the dot.
    /// If the dot is at the end of the RHS, then `dot == right.len()`.
    pub dot: usize,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum RuleIndex {
    Original(usize),
    Augmented,
}

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Lookahead {
    Terminal(DollarlessTerminalName),
    Eof,
}

#[derive(Debug, Clone)]
pub struct Transition {
    pub from: StateIndex,
    pub to: StateIndex,
    pub symbol: String,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct StateIndex(pub usize);

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Symbol {
    Terminal(DollarlessTerminalName),
    Nonterminal(String),
}
