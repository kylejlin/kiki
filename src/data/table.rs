use crate::data::*;

#[derive(Debug, Clone)]
pub struct Table {
    pub dollarless_terminals: Vec<DollarlessTerminalName>,
    pub nonterminals: Vec<String>,
    pub actions: Vec<Action>,
    pub gotos: Vec<Goto>,
}

#[derive(Debug, Clone, Copy)]
pub enum Action {
    Shift(usize),
    Reduce(usize),
    Accept,
    Err,
}

#[derive(Debug, Clone, Copy)]
pub enum Goto {
    State(usize),
    Err,
}

#[derive(Debug, Clone, Copy)]
pub enum Quasiterminal<'a> {
    Terminal(&'a DollarlessTerminalName),
    Eof,
}

impl Table {
    pub fn state_count(&self) -> usize {
        self.actions.len() / (self.dollarless_terminals.len() + 1)
    }

    /// ## Panics
    /// 1. Panics if the terminal is not in the table.
    /// 2. Panics if the state is too large.
    pub fn action(&self, state: usize, terminal: Quasiterminal) -> Action {
        let i = self.action_index(state, terminal);
        self.actions[i]
    }

    /// ## Panics
    /// 1. Panics if the terminal is not in the table.
    /// 2. Panics if the state is too large.
    pub fn set_action(&mut self, state: usize, terminal: Quasiterminal, val: Action) {
        let i = self.action_index(state, terminal);
        self.actions[i] = val;
    }

    /// ## Panics
    /// 1. Panics if the terminal is not in the table.
    /// 2. Panics if the state is too large.
    fn action_index(&self, state: usize, quasiterminal: Quasiterminal) -> usize {
        let quasiterminal_index = match quasiterminal {
            Quasiterminal::Terminal(terminal) => self
                .dollarless_terminals
                .iter()
                .position(|t| t == terminal)
                .expect("Terminal not found in table"),
            Quasiterminal::Eof => self.dollarless_terminals.len(),
        };

        if state >= self.state_count() {
            let states = self.state_count();
            panic!("State {state} is too large. There are only {states} states.");
        }

        state * (self.dollarless_terminals.len() + 1) + quasiterminal_index
    }

    /// ## Panics
    /// 1. Panics if the nonterminal is not in the table.
    /// 2. Panics if the state is too large.
    pub fn goto(&self, state: usize, nonterminal: &str) -> Goto {
        let i = self.goto_index(state, nonterminal);
        self.gotos[i]
    }

    /// ## Panics
    /// 1. Panics if the nonterminal is not in the table.
    /// 2. Panics if the state is too large.
    pub fn set_goto(&mut self, state: usize, nonterminal: &str, val: Goto) {
        let i = self.goto_index(state, nonterminal);
        self.gotos[i] = val;
    }

    /// ## Panics
    /// 1. Panics if the nonterminal is not in the table.
    /// 2. Panics if the state is too large.
    fn goto_index(&self, state: usize, nonterminal: &str) -> usize {
        let nonterminal_index = self
            .nonterminals
            .iter()
            .position(|t| t == nonterminal)
            .expect("Nonterminal not found in table");

        if state >= self.state_count() {
            let states = self.state_count();
            panic!("State {state} is too large. There are only {states} states.");
        }

        state * self.nonterminals.len() + nonterminal_index
    }
}
