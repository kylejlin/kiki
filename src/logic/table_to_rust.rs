use crate::data::{table::*, validated_file::*, KikiErr, RustSrc};
use std::collections::HashSet;

pub fn table_to_rust(table: &Table, file: ValidatedFile) -> Result<RustSrc, KikiErr> {
    let mut used_identifiers = file.defined_identifiers;
    let start_type_name = &file.start;
    let token_enum_name = &file.terminal_enum.name;
    let eof_variant_name = create_unique_identifier("Eof", &mut used_identifiers);
    let quasitoken_enum_name = create_unique_identifier("Quasitoken", &mut used_identifiers);
    let quasitoken_kind_enum_name =
        create_unique_identifier("QuasitokenKind", &mut used_identifiers);
    let nonterminal_kind_enum_name =
        create_unique_identifier("NonterminalKind", &mut used_identifiers);
    let state_enum_name = create_unique_identifier("State", &mut used_identifiers);
    let node_enum_name = create_unique_identifier("Node", &mut used_identifiers);
    let action_enum_name = create_unique_identifier("Action", &mut used_identifiers);
    let rule_kind_enum_name = create_unique_identifier("RuleKind", &mut used_identifiers);

    let token_kind_enum_variants_indent_1 = file
        .terminal_enum
        .variants
        .iter()
        .map(|variant| {
            let name = &variant.dollarless_name;
            format!("{name},\n")
        })
        .collect::<String>()
        .indent(1);

    let nonterminal_kind_enum_variants_indent_1 = file
        .nonterminals
        .iter()
        .map(|nonterminal| format!("{},\n", nonterminal.name()))
        .collect::<String>()
        .indent(1);

    let state_enum_variants_indent_1 = (0..table.states())
        .map(|i| format!("S{i},"))
        .collect::<String>()
        .indent(1);

    let node_enum_variants_indent_1: String = file
        .nonterminals
        .iter()
        .map(|nonterminal| format!("{name}({name}),", name = nonterminal.name()))
        .chain(file.terminal_enum.variants.iter().map(|variant| {
            let name = &variant.dollarless_name;
            let type_ = &variant.type_;
            format!("{name}({type_}),\n")
        }))
        .collect::<String>()
        .indent(1);

    let rule_kinds = file
        .nonterminals
        .iter()
        .map(|nonterminal| match nonterminal {
            Nonterminal::Struct(_) => 1,
            Nonterminal::Enum(e) => e.variants.len(),
        })
        .sum();
    let rule_kind_enum_variants_indent_1: String = (0..rule_kinds)
        .map(|i| format!("R{i},\n"))
        .collect::<String>()
        .indent(1);

    let pop_and_reduce_match_arms_indent_2: String = file
        .nonterminals
        .iter()
        .flat_map(|nonterminal| match nonterminal {
            Nonterminal::Struct(s) => vec![(&s.name.name, None, &s.fieldset)],
            Nonterminal::Enum(e) => e
                .variants
                .iter()
                .map(|v| (&e.name.name, Some(&v.name.name), &v.fieldset))
                .collect(),
        })
        .enumerate()
        .map(|(rule_index, (type_name, opt_variant_name, fieldset))| {
            format!(
                r#"{rule_kind_enum_name}::R{rule_index} => {{
    // TODO: Reduction code
}}"#
            )
        })
        .collect::<String>()
        .indent(2);

    let quasitoken_kind_from_token_match_arms_indent_3: String = file
        .terminal_enum
        .variants
        .iter()
        .map(|variant| {
            let name = &variant.dollarless_name;
            format!("{token_enum_name}::{name}(_) => Self::{name},\n")
        })
        .collect::<String>()
        .indent(3);

    let node_from_token_match_arms_indent_3: String = file
        .terminal_enum
        .variants
        .iter()
        .map(|variant| {
            let name = &variant.dollarless_name;
            format!("{token_enum_name}::{name}(t) => Self::{name}(t),\n")
        })
        .collect::<String>()
        .indent(3);

    let impl_try_from_node_for_each_nonterminal: String = file
        .nonterminals
        .iter()
        .map(|nonterminal| {
            let nonterminal_name = nonterminal.name();
            format!(
                r#"impl TryFrom<{node_enum_name}> for {nonterminal_name} {{
    type Error = {node_enum_name};

    fn try_from(node: {node_enum_name}) -> Result<Self, Self::Error> {{
        match node {{
            {node_enum_name}::{nonterminal_name}(n) => Ok(n),
            _ => Err(node),
        }}
    }}
}}"#
            )
        })
        .collect::<Vec<_>>()
        .join("\n\n");

    let try_into_terminal_variant_name_variant_index_fns_indent_1: String = file
        .terminal_enum
        .variants
        .iter()
        .map()
        .collect::<String>()
        .indent(1);

    Ok(RustSrc(format!(
        r#"// This code was generated by Kiki.

enum {quasitoken_enum_name} {{
    Token({token_enum_name}),
    {eof_variant_name},
}}

#[derive(Clone, Copy)]
enum {quasitoken_kind_enum_name} {{
{token_kind_enum_variants_indent_1}
    {eof_variant_name},
}}

#[derive(Clone, Copy)]
enum {nonterminal_kind_enum_name} {{
{nonterminal_kind_enum_variants_indent_1}
}}

#[derive(Clone, Copy)]
enum {state_enum_name} {{
{state_enum_variants_indent_1}
}}

enum {node_enum_name} {{
{node_enum_variants_indent_1}
}}

#[derive(Clone, Copy)]
enum {action_enum_name} {{
    Shift({state_enum_name}),
    Reduce({rule_kind_enum_name}),
    Accept,
    Err,
}}

#[derive(Clone, Copy)]
enum {rule_kind_enum_name} {{
{rule_kind_enum_variants_indent_1}
}}

pub fn parse<S>(src: S) -> Result<{start_type_name}, {token_enum_name}>
where S: IntoIterator<Item = {token_enum_name}> {{
    let mut tokens = src.into_iter()
        .map({quasitoken_enum_name}::Token)
        .chain(std::iter::once({quasitoken_enum_name}::Eof))
        .peekable();
    let mut states = vec![{state_enum_name}::S0];
    let mut nodes: Vec<{node_enum_name}> = vec![];
    loop {{
        let top_state = *states.last().unwrap();
        let next_quasitoken_kind = {quasitoken_kind_enum_name}::from_quasitoken(tokens.peek().unwrap());
        match get_action(top_state, next_quasitoken_kind) {{
            {action_enum_name}::Shift(new_state) => {{
                states.push(new_state);
                nodes.push({node_enum_name}::from_token(tokens.next().unwrap().try_into_token().unwrap()));
            }}

            {action_enum_name}::Reduce(rule_kind) => {{
                let (new_node, new_node_kind) = pop_and_reduce(&mut states, &mut nodes, rule_kind);
                nodes.push(new_node);
                let temp_top_state = *states.last().unwrap();
                let Some(new_state) = get_goto(temp_top_state, new_node_kind) else {{
                    return Err(tokens.next().unwrap());
                }};
                states.push(new_state);
            }}

            {action_enum_name}::Accept => {{
                return Ok({start_type_name}::try_from(nodes.pop().unwrap()).unwrap());
            }}

            {action_enum_name}::Err => {{
                return Err(tokens.next().unwrap());
            }}
        }}
    }}
}}

fn pop_and_reduce(states: &mut Vec<{state_enum_name}>, nodes: &mut Vec<{node_enum_name}>, rule_kind: {rule_kind_enum_name}) -> ({node_enum_name}, {nonterminal_kind_enum_name}) {{
    match rule_kind {{
{pop_and_reduce_match_arms_indent_2}
    }}
}}

impl {quasitoken_kind_enum_name} {{
    fn from_quasitoken(quasitoken: &{quasitoken_enum_name}) -> Self {{
        match quasitoken {{
            Self::Token(token) => Self::from_token(token),
            Self::{eof_variant_name} => Self::{eof_variant_name},
        }}
    }}

    fn from_token(token: &{token_enum_name}) -> Self {{
        match token {{
{quasitoken_kind_from_token_match_arms_indent_3}
        }}
    }}
}}

impl {node_enum_name} {{
    fn from_token(token: {token_enum_name}) -> Self {{
        match token {{
{node_from_token_match_arms_indent_3}
        }}
        }}
    }}
}}

impl {quasitoken_enum_name} {{
    fn try_into_token(self) -> Result<{token_enum_name}, ()> {{
        match self {{
            Self::Token(token) => Ok(token),
            Self::{eof_variant_name} => Err(()),
        }}
    }}
}}

fn get_action(top_state: {state_enum_name}, next_quasitoken_kind: {quasitoken_kind_enum_name}) -> {action_enum_name} {{
    todo!()
}}

fn get_goto(top_state: {state_enum_name}, new_node_kind: {nonterminal_kind_enum_name}) -> Option<{state_enum_name}> {{
    todo!()
}}

{impl_try_from_node_for_each_nonterminal}

impl {node_enum_name} {{
{try_into_terminal_variant_name_variant_index_fns_indent_1}
}}
"#
    )))
}

fn create_unique_identifier(preferred_name: &str, used: &mut HashSet<String>) -> String {
    if !used.contains(preferred_name) {
        used.insert(preferred_name.to_string());
        return preferred_name.to_string();
    }

    let mut i = 2;
    loop {
        let name = format!("{}{}", preferred_name, i);
        if !used.contains(&name) {
            used.insert(name.clone());
            return name;
        }
        i += 1;
    }
}

trait Indent {
    fn indent(&self, indent: usize) -> String;
}

impl Indent for str {
    fn indent(&self, level: usize) -> String {
        let mut out = String::new();
        let indent = &"    ".repeat(level);
        out.push_str(indent);
        for c in self.chars() {
            out.push(c);
            if c == '\n' {
                out.push_str(indent);
            }
        }
        out
    }
}
