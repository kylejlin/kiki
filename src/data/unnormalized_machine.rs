use crate::data::machine::{State, Transition};

use std::collections::HashSet;

#[derive(Debug, Clone)]
pub struct UnnormalizedMachine {
    /// The first state must be the start state.
    pub states: Vec<State>,
    pub transitions: HashSet<Transition>,
}
