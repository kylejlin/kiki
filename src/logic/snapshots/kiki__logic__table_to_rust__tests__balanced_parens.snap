---
source: src/logic/table_to_rust.rs
expression: rust_src
---
"// This code was generated by Kiki.\n\n#[derive(Debug)]\nenum Quasitoken {\n    Token(Token),\n    Eof,\n}\n\n#[derive(Clone, Copy, Debug)]\nenum QuasitokenKind {\n    LParen = 0,\n    RParen = 1,\n    Eof = 2,\n}\n\n#[derive(Clone, Copy, Debug)]\nenum NonterminalKind {\n    Expr = 0,\n}\n\n#[derive(Clone, Copy, Debug)]\nenum State {\n    S0 = 0,\n    S1 = 1,\n    S2 = 2,\n    S3 = 3,\n    S4 = 4,\n}\n\n#[derive(Debug)]\nenum Node {\n    Expr(Expr),\n    LParen(()),\n    RParen(()),\n}\n\n#[derive(Clone, Copy, Debug)]\nenum Action {\n    Shift(State),\n    Reduce(RuleKind),\n    Accept,\n    Err,\n}\n\n#[derive(Clone, Copy, Debug)]\nenum RuleKind {\n    R0 = 0,\n    R1 = 1,\n}\n\n/// If the parser encounters an unexpected token `t`, it will return `Err(Some(t))`.\n/// If the parser encounters an unexpected end of input, it will return `Err(None)`.\npub fn parse<S>(src: S) -> Result<Expr, Option<Token>>\nwhere S: IntoIterator<Item = Token> {\n    let mut tokens = src.into_iter()\n        .map(Quasitoken::Token)\n        .chain(std::iter::once(Quasitoken::Eof))\n        .peekable();\n    let mut states = vec![State::S0];\n    let mut nodes: Vec<Node> = vec![];\n    loop {\n        let top_state = *states.last().unwrap();\n        let next_quasitoken_kind = QuasitokenKind::from_quasitoken(tokens.peek().unwrap());\n        match get_action(top_state, next_quasitoken_kind) {\n            Action::Shift(new_state) => {\n                states.push(new_state);\n                nodes.push(Node::from_token(tokens.next().unwrap().try_into_token().unwrap()));\n            }\n\n            Action::Reduce(rule_kind) => {\n                let (new_node, new_node_kind) = pop_and_reduce(&mut states, &mut nodes, rule_kind);\n                nodes.push(new_node);\n                let temp_top_state = *states.last().unwrap();\n                let Some(new_state) = get_goto(temp_top_state, new_node_kind) else {\n                    return Err(tokens.next().unwrap().try_into_token().ok());\n                };\n                states.push(new_state);\n            }\n\n            Action::Accept => {\n                return Ok(Expr::try_from(nodes.pop().unwrap()).unwrap());\n            }\n\n            Action::Err => {\n                return Err(tokens.next().unwrap().try_into_token().ok());\n            }\n        }\n    }\n}\n\nfn pop_and_reduce(states: &mut Vec<State>, nodes: &mut Vec<Node>, rule_kind: RuleKind) -> (Node, NonterminalKind) {\n    match rule_kind {\n        RuleKind::R0 => {\n            (\n                Node::Expr(Expr::Empty),\n                NonterminalKind::Expr,\n            )\n        }\n        RuleKind::R1 => {\n            let t2 = nodes.pop().unwrap().try_into_r_paren_1().unwrap();\n            let t1 = Expr::try_from(nodes.pop().unwrap()).unwrap();\n            let t0 = nodes.pop().unwrap().try_into_l_paren_0().unwrap();\n            \n            (\n                Node::Expr(Expr::Wrap(\n                    t0,\n                    t1,\n                    t2,\n                )),\n                NonterminalKind::Expr\n            )\n        }\n    }\n}\n\nimpl QuasitokenKind {\n    fn from_quasitoken(quasitoken: &Quasitoken) -> Self {\n        match quasitoken {\n            Quasitoken::Token(token) => Self::from_token(token),\n            Quasitoken::Eof => Self::Eof,\n        }\n    }\n\n    fn from_token(token: &Token) -> Self {\n        match token {\n            Token::LParen(_) => Self::LParen,\n            Token::RParen(_) => Self::RParen,\n        }\n    }\n}\n\nimpl Node {\n    fn from_token(token: Token) -> Self {\n        match token {\n            Token::LParen(t) => Self::LParen(t),\n            Token::RParen(t) => Self::RParen(t),\n        }\n    }\n}\n\nimpl Quasitoken {\n    fn try_into_token(self) -> Result<Token, ()> {\n        match self {\n            Self::Token(token) => Ok(token),\n            Self::Eof => Err(()),\n        }\n    }\n}\n\nconst ACTION_TABLE: [[Action; 3]; 5] = [\n    [\n        Action::Shift(State::S2),\n        Action::Err,\n        Action::Reduce(RuleKind::R1),\n    ],\n    [\n        Action::Reduce(RuleKind::R1),\n        Action::Err,\n        Action::Err,\n    ],\n    [\n        Action::Err,\n        Action::Accept,\n        Action::Shift(State::S2),\n    ],\n    [\n        Action::Shift(State::S2),\n        Action::Reduce(RuleKind::R1),\n        Action::Err,\n    ],\n    [\n        Action::Err,\n        Action::Err,\n        Action::Shift(State::S4),\n    ],\n];\n\nfn get_action(top_state: State, next_quasitoken_kind: QuasitokenKind) -> Action {\n    ACTION_TABLE[top_state as usize][next_quasitoken_kind as usize]\n}\n\nconst GOTO_TABLE: [[Option<State>; 1]; 5] = [\n    [\n        Some(State::S1),\n    ],\n    [\n        None,\n    ],\n    [\n        Some(State::S3),\n    ],\n    [\n        None,\n    ],\n    [\n        None,\n    ],\n];\n\nfn get_goto(top_state: State, new_node_kind: NonterminalKind) -> Option<State> {\n    GOTO_TABLE[top_state as usize][new_node_kind as usize]\n}\n\nimpl TryFrom<Node> for Expr {\n    type Error = Node;\n\n    fn try_from(node: Node) -> Result<Self, Self::Error> {\n        match node {\n            Node::Expr(n) => Ok(n),\n            _ => Err(node),\n        }\n    }\n}\n\nimpl Node {\n    fn try_into_l_paren_0(self) -> Result<(), Self> {\n        match self {\n            Self::LParen(t) => Ok(t),\n            _ => Err(self),\n        }\n    }\n    \n    fn try_into_r_paren_1(self) -> Result<(), Self> {\n        match self {\n            Self::RParen(t) => Ok(t),\n            _ => Err(self),\n        }\n    }\n}\n\n#[derive(Clone, Debug)]\npub enum Expr {\n    Empty,\n    Wrap(\n        (),\n        Expr,\n        (),\n    ),\n}\n"
